lib/Myriad/Bootstrap.pm

```diff
@@ -78,6 +80,55 @@ sub allow_modules {
     @ALLOWED_MODULES{@_} = (1) x @_;
 }
 
+sub open_pipe {
+    # Establish comms channel for child process
+    socketpair my $child_pipe, my $parent_pipe, $constant{AF_UNIX}, $constant{SOCK_STREAM}, $constant{PF_UNSPEC}
+        or die $!;
+
+    { # Unbuffered writes
+        my $old = select($child_pipe);
+        $| = 1; select($parent_pipe);
+        $| = 1; select($old);
+    }
+
+    make_pipe_nonblocking($parent_pipe);
+    make_pipe_nonblocking($child_pipe);
+
+    return ($parent_pipe, $child_pipe);
+}
+
+sub make_pipe_nonblocking {
+    my $pipe = shift;
+    my $flags = fcntl($pipe, $constant{F_GETFL}, 0)
+        or die "Can't get flags for the socket: $!\n";
+
+    $flags = fcntl($pipe, $constant{F_SETFL}, $flags | $constant{O_NONBLOCK})
+        or die "Can't set flags for the socket: $!\n";
+}
+
+sub check_messages_in_pipe {
+    my ($pipe, $on_read) = @_;
+    # See https://perldoc.perl.org/functions/select#select-RBITS,WBITS,EBITS,TIMEOUT
+    # for a better understanding
+    my $input = '';
+    my $rin = my $win = '';
+    vec($rin, fileno($pipe), 1) = 1;
+    my $ein = $rin | $win;
+
+    die $! unless defined(my $nfound = select my $rout = $rin, my $wout = $win, my $eout = $ein, 0.5);
+
+    if($nfound) {
+        my $rslt = sysread $pipe, my $buf, 4096;
+        return 0 unless $rslt;
+        $input .= $buf;
```

Comment:
You could avoid this extra copy operation if you pass `$input` directly to `sysread` and use `length($input)` as offset.